packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart

```diff
@@ -76,90 +99,334 @@ void main() {
-      );
-
-      await dataCompleter.future;
-
-      expect(
-        service!.channel.stream,
-        emitsInOrder(
-          [
-            isA<String>().having(
-              (event) => json.decode(event),
-              'web socket stop message',
-              containsPair('type', 'stop'),
-            ),
-            isA<String>().having(
-              (event) => json.decode(event),
-              'web socket complete message',
-              containsPair('type', 'complete'),
-            ),
-          ],
-        ),
-      );
-      // bloc should disconnect due to no active subscriptions
-      expect(bloc.stream, emitsThrough(isA<DisconnectedState>()));
-
-      await streamSub.cancel();
-
-      await bloc.done.future;
+    group('should close when', () {
+      tearDown(() async {
+        bloc = null;
+        service = null; // service gets closed in  bloc
+      });
+      test('should trigger FailureState on Exception during init', () async {
+        final subscribeEvent = SubscribeEvent(
+          subscriptionRequest,
+          () {
+            service!.channel.sink.add(mockDataString);
+          },
+        );
+
+        final badService = MockWebSocketService(badInit: true);
+
+        final bloc = MockWebSocketBloc(
+          config: testApiKeyConfig,
+          authProviderRepo: getTestAuthProviderRepo(),
+          wsService: badService,
+          subscriptionOptions: subscriptionOptions,
+          pollClientOverride: mockPollClient.client,
+          connectivityOverride: Connectivity(),
+        );
+
+        expect(
+          bloc.stream,
+          emitsInOrder(
+            [
+              isA<DisconnectedState>(),
+              isA<ConnectingState>(),
+              isA<FailureState>(),
+            ],
+          ),
+        );
+
+        bloc.subscribe(
+          subscribeEvent,
+        );
+      });
+
+      test('an exception when poll responds unhealthy', () async {
+        final blocReady = Completer<void>();
+        final testDone = Completer<void>();
+        final subscribeEvent = SubscribeEvent(
+          subscriptionRequest,
+          blocReady.complete,
+        );
+        final bloc = getWebSocketBloc();
+
+        expect(
+          bloc.stream,
+          emitsInOrder(
+            [
+              isA<DisconnectedState>(),
+              isA<ConnectingState>(),
+              isA<ConnectedState>(),
+              isA<ReconnectingState>(),
+              isA<FailureState>(),
+              isA<PendingDisconnect>(),
+              isA<DisconnectedState>(),
+            ],
+          ),
+        );
+
+        bloc.subscribe(subscribeEvent).listen(
+          null,
+          onError: expectAsync1((event) {
+            testDone.complete();
+            expect(
+              event,
+              isA<ApiException>(),
+            );
+          }),
+        );
+
+        await blocReady.future;
+        mockPollClient.sendUnhealthyResponse = true;
+        await testDone.future;
```

Comment:
This shouldn't be needed - expectAsync1 will prevent the test from ending, or should at least.